contracts/src/core/BatcherPaymentService.sol

```diff
@@ -125,4 +144,75 @@ contract BatcherPaymentService is
         );
         _;
     }
+
+    function checkMerkleRootAndVerifySignatures(
+        bytes32[] calldata leaves,
+        bytes32 batchMerkleRoot,
+        SignatureData[] calldata signatures,
+        uint256 feePerProof
+    ) public {
+        uint256 numNodesInLayer = leaves.length / 2;
+        bytes32[] memory layer = new bytes32[](numNodesInLayer);
+
+        uint32 i = 0;
+
+        // Calculate the hash of the next layer of the Merkle tree
+        // and verify the signatures up to numNodesInLayer
+        for (i = 0; i < numNodesInLayer; i++) {
+            layer[i] = keccak256(
+                abi.encodePacked(leaves[2 * i], leaves[2 * i + 1])
+            );
+
+            verifySignatureAndNonce(leaves[i], signatures[i], feePerProof);
+        }
+
+        // Verify the rest of the signatures
+        for (; i < signatures.length; i++) {
+            verifySignatureAndNonce(leaves[i], signatures[i], feePerProof);
+        }
+
+        // The next layer above has half as many nodes
+        numNodesInLayer /= 2;
+
+        // Continue calculating Merkle root for remaining layers
+        while (numNodesInLayer != 0) {
+            // Overwrite the first numNodesInLayer nodes in layer with the pairwise hashes of their children
+            for (i = 0; i < numNodesInLayer; i++) {
+                layer[i] = keccak256(
+                    abi.encodePacked(layer[2 * i], layer[2 * i + 1])
+                );
+            }
+
+            // The next layer above has half as many nodes
+            numNodesInLayer /= 2;
+        }
+
+        require(layer[0] == batchMerkleRoot, "Invalid merkle root");
+    }
+
+    function verifySignatureAndNonce(
```

Comment:
small nit, it may be good if the function has something related to the balances in its name, since it is decreasing the balances after verifying